### ISSUE TITLE
Array Types (v1)

### ISSUE THREAD

**Issue by thevirtuoso1973** [modus-continens/modus#180]
Initial implementation of a new type - arrays. Should satisfy the needs of #71.

**Comment by thevirtuoso1973** [comment]
So it does appear that the implementation I'm currently working on will end up treating non-matching array sizes as failing to unify and therefore a resolution error.
I think we'd need some way of declaring types in head literals, for example, for better error messages.

**Comment by mechtaev** [comment]
We also need to introduce predicate types, where arguments of predicates can be either strings or arrays of strings. We will also assume that we don't have an polymorphic predicates for now, this is why = is going to be only used for strings.

**Comment by mechtaev** [comment]
As for non-matching arrays sizes, why is it an issue? In any case, if we only use arrays as arguments of builtin predicates, we will only unify arrays with variables, so there will be no size comparision.

**Comment by thevirtuoso1973** [comment]
> if we only use arrays as arguments of builtin predicates

Yes this is not an issue if we only allow arrays in built-in predicates.
